### ISSUE TITLE
Run GLM 4.7 and deepseek on BH Galaxy

### ISSUE THREAD

**Issue by AleksKnezevic** [tenstorrent/tt-xla#3927]

**Comment by btsoiTT** [comment]
Updates on progress for DeepSeek V3.1 / GLM-4.7 Galaxy end-to-end work

Branch [btsoi/deepseek_glm_galaxy_e2e_cleaned](https://github.com/tenstorrent/tt-xla/tree/btsoi/deepseek_glm_galaxy_e2e_cleaned) ships four tests - a prefill-only `*_full_sparse_moe` and an autoregressive `*_decode_static_cache` for each of DeepSeek V3.1 and GLM-4.7. All four share: batch 32, 32-token prompt, Galaxy (4,8) mesh.

## Prerequisites

1. Access to a Galaxy machine.
2. Confirm there is enough disk storage for the weights and swap
3. Set up swap. The full-layer DeepSeek V3.1 / GLM-4.7 runs peak well above system RAM during weight dequant + expert stacking, so a large swapfile is required to avoid OOM (1.5TB is probably enough, but more doesn't hurt):

   ```bash
   # To enable swap

   # only if the file system is btrfs
   sudo truncate -s 0 ./swapfile
   sudo chattr +C ./swapfile

   sudo fallocate -l 3TB ./swapfile
   sudo chmod 600 ./swapfile
   sudo mkswap ./swapfile
   sudo swapon ./swapfile


   # To disable swap

   sudo swapoff ./swapfile
   sudo rm ./swapfile
   ```

Note that this needs to be run outside of the container.

## Weight caches

To reduce runtime peak memory usage, a weight caching mechanism is implemented.  Both tests can load weights from a pre-built on-disk cache instead of re-running the heavy preprocessing each launch. There are two flavors:

- **Dequant cache** - DeepSeek weights ship as FP8 on HF but the model runs in BF16; dequantizing the full checkpoint takes a long time and the FP8 weights stick around which increases memory pressure, so we do it once and cache the BF16 result. GLM-4.7 is already BF16, so no dequant step is needed.
- **Post-sparse cache** - To run both models, we call `enable_sparse_mlp` to use `sparse_matmul`. The function stacks MoE experts as a single `[E, in, out]` tensor. The post- sparse cache caches this tensor to further reduce the memory consumption and mmap it to reduce disk write pressure during swapping.

Caches land under `$HF_HOME/tt_xla_dequant_cache/...` (`_post_sparse` suffix for the stacked variant). To build them, run the following (this may take 1-2 hours):

```bash
# DeepSeek V3.1 - FP8 → BF16 dequant, then post-sparse stacking
cd tests/torch/models/deepseek_v3_2_exp
python build_weight_cache.py --n-layers 61
python build_weight_cache.py --n-layers 61 --post-sparse

# GLM-4.7 - post-sparse stacking only (weights are already BF16)
cd tests/torch/models/glm4_moe
python build_weight_cache_glm.py --n-layers 92
```

Note that right now, layer n gets its own cache, which means the weights get duplicated. For example, 
running tests with `GLM_N_LAYERS=92` requires
```
# builds `~/.cache/huggingface/tt_xla_dequant_cache/zai-org--GLM-4.7_92layers_post_sparse/`
python build_weight_cache_glm.py --n-layers 92  
``` 
but running with `GLM_N_LAYERS=4` requires 
```
# builds `~/.cache/huggingface/tt_xla_dequant_cache/zai-org--GLM-4.7_4layers_post_sparse/`
python build_weight_cache_glm.py --n-layers 4
``` 
so the weights of the 4 layers are duplicated. 
This is an implementation flaw and should be fixed. The 4 layer run should be able to reuse the cached weights in the 61 layer cache. 


## Decode tests

Run either decode test from the repo root. Layer count must match the cache you built; set `*_CPU_REFERENCE=1` to also run a CPU golden and get per-step PCC (Note that for many layers, CPU can be slow, taking an extra 30 minutes to 2 hours):

```bash
# DeepSeek - full 61 layers, TT only, expected to take around 45 minutes
DEEPSEEK_N_LAYERS=61 pytest -svv \
  tests/torch/models/deepseek_v3_2_exp/test_deepseek_v3_1.py::test_deepseek_v3_1_decode_static_cache

# DeepSeek - 4 layers, with CPU reference & per-step PCC
DEEPSEEK_N_LAYERS=4 DEEPSEEK_CPU_REFERENCE=1 pytest -svv \
  tests/torch/models/deepseek_v3_2_exp/test_deepseek_v3_1.py::test_deepseek_v3_1_decode_static_cache

# GLM - full 92 layers, TT only, expected to take around 30 minutes
GLM_N_LAYERS=92 pytest -svv \
  tests/torch/models/glm4_moe/test_glm4_7.py::test_glm4_7_decode_static_cache

# GLM - 4 layers
GLM_N_LAYERS=4 GLM_CPU_REFERENCE=1 pytest -svv tests/torch/models/glm4_moe/test_glm4_7.py::test_glm4_7_decode_static_cache

```

Both tests print timing for model load, move-to-device, prefill compile+run, each decode step and decoded top-1 tokens.

## `*_full_sparse_moe` tests

Each decode test has a prefill-only counterpart (`*_full_sparse_moe`) that runs one `torch.compile` trace and compares TT vs CPU logits - effectively a subset of the decode test, so it's not strictly necessary once decode works. Uses the same weight caches as the decode tests; no extra build step.

Env vars match the decode variants:

```bash
DEEPSEEK_N_LAYERS=4 DEEPSEEK_CPU_REFERENCE=1 pytest -svv \
  tests/torch/models/deepseek_v3_2_exp/test_deepseek_v3_1.py::test_deepseek_v3_1_full_sparse_moe

GLM_N_LAYERS=4 GLM_CPU_REFERENCE=1 pytest -svv \
  tests/torch/models/glm4_moe/test_glm4_7.py::test_glm4_7_full_sparse_moe
```

The `*_full_sparse_moe` tests are the same pattern as `tests/infra/testers/single_chip/graph/graph_tester.py::run_graph_test`. The reason we don't call `run_graph_test` is because we want some custom instrumentation (per-step timing, top-k token prints etc) for debugging.

**PCC vs coherent tokens.** PCC > ~0.95 is considered good, but we've seen sub-0.95 runs still emit coherent text, and high first-step PCC doesn't guarantee coherence deep into decode. For bring-up, the primary check is eyeballing `[TT full]` / `[TT tokens]` for coherent output; treat PCC as a diagnostic, not a gate.

## Suggested Next Steps

- Bring up Kimi K2.5 with experts in bf4
- Bring up DeepSeek V3.2 (Needs an optimized MLA decode implementation - reach out to Het for the branch) 
- Refactor dequant + sparse cache logic into the test files. Today each model ships a separate `build_weight_cache*.py` script that must be run manually before the pytest. Fold that logic into the test setup (build on first run, reuse on subsequent runs) so the tests are self-contained and a single `pytest` invocation is enough.
- Share one cache across layer counts 
- Refactor tests to fit repo conventions. `*_full_sparse_moe` should be rewritten on top of `run_graph_test`, and the decode tests should live as standalone programs under `examples/pytorch/` rather than as pytests.

**Comment by gengelageTT** [comment]
4 layer benchmarks (with proper weights) have been added for both of these models and adding full model tests is now being tracked by https://github.com/tenstorrent/tt-xla/issues/4822

**Comment by gengelageTT** [comment]
Actually I might be misunderstanding that issue. Going to leave this open for now
